pylops/signalprocessing/dtcwt.py

```diff
@@ -0,0 +1,145 @@
+    -----
+    The :py:func:`dtcwt` library uses a Pyramid object to represent the transform domain signal.
+    It has
+        `lowpass` (coarsest scale lowpass signal)
+        `highpasses` (complex subband coefficients for corresponding scales)
+        `scales` (lowpass signal for corresponding scales finest to coarsest)
+
+    To make the dtcwt forward() and inverse() functions compatible with pylops, the Pyramid object is
+    flattened out and all coeffs(highpasses and low pass signal) are appened into one array using the
+    `_coeff_to_array` method.
+    For inverse, the flattened array is used to reconstruct the Pyramid object using the `_array_to_coeff`
+    method and then inverse is performed.
+
+    """
+
+    def __init__(
+        self,
+        dims: Union[int, InputDimsLike],
+        transform: int = 1,
+        biort: str = "near_sym_a",
+        qshift: str = "qshift_a",
+        nlevels: int = 3,
+        include_scale: bool = False,
+        dtype: DTypeLike = "float64",
+        name: str = "C",
+    ) -> None:
+        self.dims = _value_or_sized_to_tuple(dims)
+        self.transform = transform
+        self.ndim = len(self.dims)
+        self.nlevels = nlevels
+        self.include_scale = include_scale
+
+        if self.transform == 1:
+            self._transform = dtcwt.Transform1d(biort=biort, qshift=qshift)
+        elif self.transform == 2:
+            raise NotImplementedError("DTCWT is not implmented for 2D")
+        elif self.transform == 3:
+            raise NotImplementedError("DTCWT is not implmented for 3D")
+        else:
+            raise ValueError("DTCWT only supports 1D, 2D and 3D")
+
+        pyr = self._transform.forward(
+            np.ones(self.dims), nlevels=self.nlevels, include_scale=True
+        )
+        self.coeff_array_size = 0
+        self.lowpass_size = len(pyr.lowpass)
+        self.slices = []
+        for _h in pyr.highpasses:
+            self.slices.append(len(_h))
+            self.coeff_array_size += len(_h)
+        self.coeff_array_size += self.lowpass_size
+        self.second_dim = 1
+        if len(dims) > 1:
+            self.coeff_array_size *= self.dims[1]
+            self.lowpass_size *= self.dims[1]
+            self.second_dim = self.dims[1]
+        super().__init__(
+            dtype=np.dtype(dtype),
+            dims=self.dims,
+            dimsd=(self.coeff_array_size,),
+            name=name,
+        )
+
+    def _coeff_to_array(self, pyr: dtcwt.Pyramid) -> NDArray:
+        print("og lowpass ", pyr.lowpass)
+        print("og highpasses ", pyr.highpasses)
+        coeffs = pyr.highpasses
+        flat_coeffs = []
+        for band in coeffs:
+            for c in band:
+                flat_coeffs.append(c)
+        flat_coeffs = np.concatenate((flat_coeffs, pyr.lowpass))
+        return flat_coeffs
+
+    def _array_to_coeff(self, X: NDArray) -> dtcwt.Pyramid:
+        lowpass = np.array([x.real for x in X[-1 * self.lowpass_size :]]).reshape(
```

Comment:
why not just `-self.lowpass_size`

Reply:
updated in [f887031](https://github.com/PyLops/pylops/pull/495/commits/f887031ff0aed0edac70bed9bf1388e2dbfb4b70)